Source/CountDown.spoon/init.lua

```diff
@@ -1,85 +1,638 @@
+   local hours =   math.floor(obj.timeLeft / 3600)
+   local seconds = obj.timeLeft % 60 
+   local timeString
+   if hours > 0 then
+      timerString = string.format("%s %d:%02d:%02d", obj.currentIcon, hours, minutes, seconds)
+   else
+      timerString = string.format("%s %02d:%02d", obj.currentIcon, minutes, seconds)
+   end
+   obj.menuBar:setTitle(timerString)
+end
+
+function obj:menuBar_update(isPaused)
+   -- update the icons and textual information in menu bar
+   -- if active, indicate time left
+   if not obj.menuBar:isInMenuBar() then
+      obj.menuBar:returnToMenuBar()
+   end
+   local label = nil
+   if isPaused then
+      label = string.format("%s Resume", obj.menuBarIconActive)
+      obj.currentIcon = obj.menuBarIconPause
+   else
+      obj.currentIcon = obj.menuBarIconActive
+      label = string.format("%s Pause", obj.menuBarIconPause)
+   end
+   local items = {
+      { title = string.format("%s Stop", obj.menuBarIconStop),
+         fn = function() obj:cancel() end },
+      { title = label, fn = function() obj:pauseOrResume() end }
+   }
+   obj.menuBar:setMenu(items)
+   obj:menuBar_updateTimerString()
+end
+
+--- callback
+
+function obj:tick()
+   -- main timer callback
+   -- keeps track of the time passed
+   -- ticks every second...
+   --   updates barProgress
+   --   and potentially issues a warning
+   if not obj.timerRunning then
+      print("This should not happen!!")
+      return
+   end
+   local timeNow = obj:time_absolute_seconds()
+   obj.timeLeft = obj.endingTime - timeNow
+   if obj.timeLeft <= 0 then
+      obj.timerRunning = false
+      -- we need to save this before we reset
+      -- and we need to reset before we notify
+      local requestedMinutes = obj.timerLenMinutes
+      obj:reset_timer()
+      obj:end_of_timer_notify(requestedMinutes)
+   else --update progress ratio
+      obj.barProgress = 1 - obj.timeLeft/(obj.timerLenMinutes * 60)
+      obj:warning_update()
+      obj:menuBar_updateTimerString()
+   end
+end
+
+function obj:create_tick_timer()
+   obj.endingTime = obj:time_absolute_seconds() + obj.timerLenMinutes * 60
+   obj.timerRunning = true
+   obj:tick()
+   obj.timer = hs.timer.doWhile(function() return obj.timerRunning end,
+      function() obj:tick() end,
+      1)
+end
+-------------------------------
+--- API
+
+
 --- CountDown:startFor(minutes)
 --- Method
 --- Start a countdown for `minutes` minutes immediately. Calling this method again will kill the existing countdown instance.
 ---
 --- Parameters:
 ---  * minutes - How many minutes
+---              Defaults to obj.defaultLenMinutes
+---  * callback: optional: a function to call when the timer 
+---            is up. it takes one parameters (minutes)
+---            the minutes that were requested.
+---            The callback is not called if timer is cancelled
+function obj:startFor(minutes, callback)
+   if not minutes  then
+      minutes = obj.defaultLenMinutes
+   end
+   if obj.timerRunning then
```

Comment:
maybe do here a timer:cancel() and start over with a new timer?

Reply:
I think it is better to issue an error. The reasons is that one might accidentally try to set another timer without noticing the current one is running. I guess a better option would be to make this optional (via a variable). What do you think?

Reply:
or alternatively, ask the user.

Reply:
I mean that if minutes would be an optional parameter, but someone want a callback 
`function obj:startFor([minutes,] callback)`

I have seen this on some places at hammerspoon:

```lua
if type(minutes)=='function' then
	callback = minutes
	minutes = obj.defaultLenMinutes
end
```
example:
https://github.com/Hammerspoon/hammerspoon/blob/ec9a8ead1fc9074b8647c5ee4c09ba96f66a719a/extensions/host/host.lua#L89-L95